### PR TITLE
[Merged by Bors] - Update node health endpoint

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

[#4292](https://github.com/sigp/lighthouse/issues/4292)

## Proposed Changes

Updated the node health endpoint

will return a 200 status code if  `!syncing && !el_offline && !optimistic`

wil return a 206 if `(syncing || optimistic) &&  !el_offline`

will return a 503 if `el_offline`

### CODE REVIEW

File: beacon_node/http_api/src/lib.rs [hunk 1]
```diff
@@ -2328,24 +2328,39 @@ pub fn serve<T: BeaconChainTypes>(
...
+                    let is_optimistic = chain
+                        .is_optimistic_or_invalid_head()
+                        .map_err(warp_utils::reject::beacon_chain_error)?;
+
+                    let is_syncing = network_globals.sync_state.read().is_syncing();
+
+                    let unhealthy = is_syncing || Some(is_optimistic) || Some(el_offline);
+
+                    if (unhealthy) {
+                        Err(warp_utils::reject::not_synced(format!(
```
Comment (eserilev) [inline beacon_node/http_api/src/lib.rs]: could have returned a different error message for each situation where unhealthy resolves to true, but im not sure how helpful that is

**Comment by eserilev** [comment]
hopefully im on the right track here

**Comment by eserilev** [comment]
`test_get_node_health` is failing. will need to figure out how to get that test to pass (assuming my implementation is correct)

**Comment by paulhauner** [comment]
Thanks for your contribution. I'd like to note that the [Beacon-API specs](https://ethereum.github.io/beacon-APIs/#/Node/getHealth) state that:

> 206    Node is syncing but can serve incomplete data

So, moving to a `NotSynced` (503) would be going against the spec. I'm also not sure it's clear that syncing node is necessarily unhealthy. For example, a newly provisioned node that is syncing is "healthy" in the sense that it's working as intended and requires no intervention.

Fine-grained information is available at [`node/syncing`](https://ethereum.github.io/beacon-APIs/#/Node/getSyncingStatus).

I do agree that a node with a `ee_offline` is unhealthy, though. I'd be open to seeing an error returned when `ee_offline == true`.

**Comment by michaelsproul** [comment]
Yeah my bad, I didn't distinguish between syncing and EL offline in the original issue. I agree we should return a 206 for syncing/optimistic, and a 503 for `el_offline`.

**Comment by eserilev** [comment]
thanks for the feedback!

File: beacon_node/http_api/tests/status_tests.rs [hunk 2]
```diff
@@ -143,3 +144,75 @@ async fn el_error_on_new_payload() {
...
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+async fn node_health_el_online_and_not_synced() {
+    let num_blocks = E::slots_per_epoch() / 2;
+    let num_validators = E::slots_per_epoch();
+    let tester = post_merge_tester(num_blocks, num_validators).await;
+    let harness = &tester.harness;
+    let mock_el = harness.mock_execution_layer.as_ref().unwrap();
+
+    // EL not synced
+    mock_el.server.set_syncing_response(Ok(true));
```
Comment (eserilev) [inline beacon_node/http_api/tests/status_tests.rs]: This test currently fails. I'm trying to test the situation where the node is syncing and the endpoint returns a 206 status code. However i'm not really sure how to mock `syncing` to true

Comment (eserilev) [reply beacon_node/http_api/tests/status_tests.rs]: @michaelsproul ideally i'd like to test the situation where the node health endpoint returns a 206, but I'm struggling trying to find a way to mock the situation where the node is syncing/optimistic

Comment (michaelsproul) [reply beacon_node/http_api/tests/status_tests.rs]: You can call `mock_el.server.all_payloads_syncing(true)`, then add a block (e.g. with `extend_chain`). That new block will result in a SYNCING status from the EL on `newPayload`, which will put Lighthouse into optimistic sync

**Comment by eserilev** [comment]
@michaelsproul thanks for the feedback and help. this should be ready for another review

File: beacon_node/http_api/src/lib.rs [hunk 3]
```diff
@@ -2333,24 +2333,43 @@ pub fn serve<T: BeaconChainTypes>(
...
-                SyncState::Stalled => Err(warp_utils::reject::not_synced(
-                    "sync stalled, beacon chain may not yet be initialized.".to_string(),
-                )),
-            })
-        });
+        .and(chain_filter.clone())
+        .and_then(
+            |network_globals: Arc<NetworkGlobals<T::EthSpec>>, chain: Arc<BeaconChain<T>>| {
+                async move {
+
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]:
```suggestion
```

File: beacon_node/http_api/src/lib.rs [hunk 4]
```diff
@@ -2333,24 +2333,43 @@ pub fn serve<T: BeaconChainTypes>(
...
+                    let el_offline = if let Some(el) = &chain.execution_layer {
+                        el.is_offline_or_erroring().await
+                    } else {
+                        true
+                    };
+
+                    blocking_response_task( move || {
+
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]:
```suggestion
                    blocking_response_task(move || {
```

File: beacon_node/http_api/src/lib.rs [hunk 5]
```diff
@@ -2333,24 +2333,43 @@ pub fn serve<T: BeaconChainTypes>(
...
+                        let is_optimistic = chain
+                            .is_optimistic_or_invalid_head()
+                            .map_err(warp_utils::reject::beacon_chain_error)?;
+
+                        let is_syncing = network_globals.sync_state.read().is_syncing();
```
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: I realised we lost the previous behaviour for `Sync::Stalled`. Maybe as a simple solution we could change this to:

```suggestion
                        let is_syncing = !network_globals.sync_state.read().is_synced();
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]

File: beacon_node/http_api/src/lib.rs (hunk 5 above)
Comment (eserilev) [reply beacon_node/http_api/src/lib.rs]: ah yep, ill make that change. thanks

**Review by michaelsproul — APPROVED** [review]
Nice, thanks!

**Pull request closed without merging**